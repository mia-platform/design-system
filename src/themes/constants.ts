/* eslint-disable no-underscore-dangle */
import { dirname, resolve } from 'path'
import { fileURLToPath } from 'url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

export const THEMES_DIR = resolve(__dirname, 'files')
export const THEME_GENERATOR_FILE = 'theme-generator.json'
export const PRIMITIVES_FILE = 'primitives.json'
export const GENERATED_FILE = 'theme.json'
