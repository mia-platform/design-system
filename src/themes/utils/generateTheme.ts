/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

/* eslint-disable array-callback-return */
/* eslint-disable func-names */

import { readFileSync, writeFileSync } from 'fs'
import get from 'lodash/get'
import { resolve } from 'path'
import traverse from 'traverse'

import Theme from '../schema'

export const THEMES_DIR = resolve(__dirname, '../files')
export const THEME_GENERATOR_FILE = 'theme-generator.json'
export const GLOBAL_FILE = 'global.json'
export const GENERATED_FILE = 'theme.json'

const INTERPOLATED_VALUE = /^{.*}$/g
const PARENTHESES = /[{}]/g

const getFile = (theme: string, file: string): string => resolve(THEMES_DIR, theme, file)

/**
 * Resolves interpolated values from a global definitions file in the theme.
 *
 * @param {Theme} themeValues - The theme values to resolve interpolated values from.
 * @returns {func} A function that resolves interpolated values.
 */
const resolveThemeValues = (themeValues: Theme) => (node: string) => {
  if (node.match?.(INTERPOLATED_VALUE)) {
    const path = node.replace(PARENTHESES, '').split('.')

    const { value } = get(themeValues, path) || {}

    return value
  }
  return node
}

/**
 * Generates a theme based on the provided theme name.
 *
 * @param {string} themeName - The name of the theme to generate.
 * @returns {Promise<void>} A promise that resolves when the theme generation is complete.
 */
export default async function generateTheme(themeName: string): Promise<void> {
  const structure = await readFileSync(getFile(themeName, THEME_GENERATOR_FILE)).toString()
  const values = await readFileSync(getFile(themeName, GLOBAL_FILE)).toString()

  const themeStructure: object = JSON.parse(structure)
  const themeValues: Theme = JSON.parse(values)

  const resolveVariable = resolveThemeValues(themeValues)

  const resolvedTheme: Theme = traverse(themeStructure).map(function(node) {
    const isLeaf = node?.value && node?.type
    const resolvedNode = resolveVariable(isLeaf ? node.value : node)

    this.update(resolvedNode)
  })

  writeFileSync(getFile(themeName, GENERATED_FILE), JSON.stringify(resolvedTheme, null, 2))
}
