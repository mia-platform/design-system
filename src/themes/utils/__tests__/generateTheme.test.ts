/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { readFileSync, readdirSync } from 'fs'
import { resolve } from 'path'

import { GENERATED_FILE, THEMES_DIR } from '../../constants'
import { generateTheme } from '../generateTheme'


describe('Generate Theme', () => {
  for (const themeName of readdirSync(THEMES_DIR)) {
    test(`${themeName} matches snapshot`, async() => {
      const generated = readFileSync(resolve(THEMES_DIR, themeName, GENERATED_FILE)).toString()
      expect(JSON.parse(generated)).toMatchSnapshot()
    })
  }

  test('throws error if there is some error on theme generation file', async() => {
    const themeGeneratorFilePath = resolve(__dirname, '../mocks/theme-generator-with-error.json')
    const primitivesFilePath = resolve(__dirname, '../mocks/primitives.json')

    expect(() => generateTheme(themeGeneratorFilePath, primitivesFilePath)).toThrow('Something went wrong resolving {this.primitive.not.exists}')
  })
})
