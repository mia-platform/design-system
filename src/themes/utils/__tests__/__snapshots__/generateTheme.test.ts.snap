// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Theme generates dark theme configuration 1`] = `
{
  "palette": {
    "action": {
      "danger": {
        "active": "#f5222d",
        "disabled": "#f5f5f5",
        "focus": "#f5222d",
        "hover": "#ff4d4f",
        "selected": "#d9363e",
      },
      "primary": {
        "active": "#1890ff",
        "disabled": "#f5f5f5",
        "focus": "#40a9ff",
        "hover": "#40a9ff",
        "selected": "#096dd9",
      },
      "secondary": {
        "active": "#ffffff",
        "disabled": "#f5f5f5",
        "focus": "#40a9ff",
        "hover": "#40a9ff",
        "selected": "#096dd9",
      },
    },
    "background": {
      "0": "#ffffff",
      "100": "#ffffff",
      "200": "#ffffff",
      "300": "#ffffff",
      "400": "#ffffff",
      "500": "#ffffff",
    },
    "common": {
      "black": "#000000",
      "grey": {
        "100": "#f5f5f5",
        "200": "#f0f0f0",
        "300": "#e6e6e6",
        "400": "#d9d9d9",
        "50": "#fafafa",
        "500": "#d4d4d4",
        "600": "#aaaaaa",
        "700": "#858585",
        "800": "#666666",
        "900": "#444444",
      },
      "white": "#ffffff",
    },
    "error": {
      "contrastText": "#ffffff",
      "dark": "#d9363e",
      "light": "#ff4d4f",
      "main": "#f5222d",
    },
    "info": {
      "contrastText": "#666666",
      "dark": "#1890ff",
      "light": "#e6f7ff",
      "main": "#9cdcfe",
    },
    "link": {
      "active": "#1890ff",
      "disabled": "#f5f5f5",
      "focus": "#096dd9",
      "hover": "#40a9ff",
      "selected": "#096dd9",
    },
    "mode": "dark",
    "primary": {
      "contrastText": "#ffffff",
      "dark": "#096dd9",
      "light": "#40a9ff",
      "main": "#1890ff",
    },
    "secondary": {
      "contrastText": "#666666",
      "dark": "#666666",
      "light": "#d9d9d9",
      "main": "#aaaaaa",
    },
    "success": {
      "contrastText": "#666666",
      "dark": "#52c41a",
      "light": "#f6ffed",
      "main": "#b7eb8f",
    },
    "warning": {
      "contrastText": "#666666",
      "dark": "#ff9933",
      "light": "#fffbe6",
      "main": "#ffe58f",
    },
  },
  "shape": {
    "border": {
      "radius": {
        "lg": "8",
        "md": "4",
        "sm": "2",
        "xl": "16",
        "xs": "1",
      },
      "width": {
        "lg": "8",
        "md": "4",
        "sm": "2",
        "xl": "16",
        "xs": "1",
      },
    },
    "size": {
      "2xl": "24",
      "3xl": "32",
      "4xl": "64",
      "5xl": "96",
      "lg": "8",
      "md": "4",
      "sm": "2",
      "xl": "16",
      "xs": "1",
    },
  },
  "spacing": {
    "gap": {
      "lg": "24",
      "md": "16",
      "sm": "8",
      "xl": "32",
      "xs": "4",
    },
    "margin": {
      "lg": "24",
      "md": "16",
      "sm": "8",
      "xl": "32",
      "xs": "4",
    },
    "padding": {
      "lg": "24",
      "md": "16",
      "sm": "8",
      "xl": "32",
      "xs": "4",
    },
  },
  "typography": {
    "action": {
      "fontFamily": "Inter",
      "fontSize": "16",
      "fontWeight": "400",
      "lineHeight": "16",
    },
    "bodyL": {
      "fontFamily": "Inter",
      "fontSize": "18",
      "fontWeight": "400",
      "lineHeight": "24",
    },
    "bodyLBold": {
      "fontFamily": "Inter",
      "fontSize": "18",
      "fontWeight": "600",
      "lineHeight": "24",
    },
    "bodyM": {
      "fontFamily": "Inter",
      "fontSize": "16",
      "fontWeight": "400",
      "lineHeight": "24",
    },
    "bodyMBold": {
      "fontFamily": "Inter",
      "fontSize": "16",
      "fontWeight": "600",
      "lineHeight": "24",
    },
    "bodyS": {
      "fontFamily": "Inter",
      "fontSize": "14",
      "fontWeight": "400",
      "lineHeight": "16",
    },
    "bodySBold": {
      "fontFamily": "Inter",
      "fontSize": "14",
      "fontWeight": "600",
      "lineHeight": "16",
    },
    "caption": {
      "fontFamily": "Inter",
      "fontSize": "12",
      "fontWeight": "400",
      "lineHeight": "12",
    },
    "code": {
      "fontFamily": "Menlo",
      "fontSize": "14",
      "fontWeight": "400",
      "lineHeight": "24",
    },
    "h1": {
      "fontFamily": "Inter",
      "fontSize": "24",
      "fontWeight": "600",
      "lineHeight": "32",
    },
    "h2": {
      "fontFamily": "Inter",
      "fontSize": "20",
      "fontWeight": "600",
      "lineHeight": "32",
    },
    "h3": {
      "fontFamily": "Inter",
      "fontSize": "18",
      "fontWeight": "600",
      "lineHeight": "24",
    },
    "h4": {
      "fontFamily": "Inter",
      "fontSize": "16",
      "fontWeight": "600",
      "lineHeight": "24",
    },
    "link": {
      "fontFamily": "Inter",
      "fontSize": "16",
      "fontWeight": "400",
      "lineHeight": "24",
      "textDecoration": "underline",
    },
  },
}
`;

exports[`Generate Theme generates light theme configuration 1`] = `
{
  "palette": {
    "action": {
      "danger": {
        "active": "#f5222d",
        "disabled": "#f5f5f5",
        "focus": "#f5222d",
        "hover": "#ff4d4f",
        "selected": "#d9363e",
      },
      "primary": {
        "active": "#1890ff",
        "disabled": "#f5f5f5",
        "focus": "#40a9ff",
        "hover": "#40a9ff",
        "selected": "#096dd9",
      },
      "secondary": {
        "active": "#ffffff",
        "disabled": "#f5f5f5",
        "focus": "#40a9ff",
        "hover": "#40a9ff",
        "selected": "#096dd9",
      },
    },
    "background": {
      "0": "#ffffff",
      "100": "#ffffff",
      "200": "#ffffff",
      "300": "#ffffff",
      "400": "#ffffff",
      "500": "#ffffff",
    },
    "common": {
      "black": "#000000",
      "grey": {
        "100": "#f5f5f5",
        "200": "#f0f0f0",
        "300": "#e6e6e6",
        "400": "#d9d9d9",
        "50": "#fafafa",
        "500": "#d4d4d4",
        "600": "#aaaaaa",
        "700": "#858585",
        "800": "#666666",
        "900": "#444444",
      },
      "white": "#ffffff",
    },
    "error": {
      "contrastText": "#ffffff",
      "dark": "#d9363e",
      "light": "#ff4d4f",
      "main": "#f5222d",
    },
    "info": {
      "contrastText": "#666666",
      "dark": "#1890ff",
      "light": "#e6f7ff",
      "main": "#9cdcfe",
    },
    "link": {
      "active": "#1890ff",
      "disabled": "#f5f5f5",
      "focus": "#096dd9",
      "hover": "#40a9ff",
      "selected": "#096dd9",
    },
    "mode": "light",
    "primary": {
      "contrastText": "#ffffff",
      "dark": "#096dd9",
      "light": "#40a9ff",
      "main": "#1890ff",
    },
    "secondary": {
      "contrastText": "#666666",
      "dark": "#666666",
      "light": "#d9d9d9",
      "main": "#aaaaaa",
    },
    "success": {
      "contrastText": "#666666",
      "dark": "#52c41a",
      "light": "#f6ffed",
      "main": "#b7eb8f",
    },
    "warning": {
      "contrastText": "#666666",
      "dark": "#ff9933",
      "light": "#fffbe6",
      "main": "#ffe58f",
    },
  },
  "shape": {
    "border": {
      "radius": {
        "lg": "8",
        "md": "4",
        "sm": "2",
        "xl": "16",
        "xs": "1",
      },
      "width": {
        "lg": "8",
        "md": "4",
        "sm": "2",
        "xl": "16",
        "xs": "1",
      },
    },
    "size": {
      "2xl": "24",
      "3xl": "32",
      "4xl": "48",
      "5xl": "64",
      "6xl": "96",
      "lg": "8",
      "md": "4",
      "sm": "2",
      "xl": "16",
      "xs": "1",
    },
  },
  "spacing": {
    "gap": {
      "lg": "24",
      "md": "16",
      "sm": "8",
      "xl": "32",
      "xs": "4",
    },
    "margin": {
      "lg": "24",
      "md": "16",
      "sm": "8",
      "xl": "32",
      "xs": "4",
    },
    "padding": {
      "lg": "24",
      "md": "16",
      "sm": "8",
      "xl": "32",
      "xs": "4",
    },
  },
  "typography": {
    "action": {
      "fontFamily": "Inter",
      "fontSize": "14",
      "fontWeight": "400",
      "lineHeight": "16",
    },
    "bodyL": {
      "fontFamily": "Inter",
      "fontSize": "18",
      "fontWeight": "400",
      "lineHeight": "24",
    },
    "bodyLBold": {
      "fontFamily": "Inter",
      "fontSize": "18",
      "fontWeight": "600",
      "lineHeight": "24",
    },
    "bodyM": {
      "fontFamily": "Inter",
      "fontSize": "16",
      "fontWeight": "400",
      "lineHeight": "24",
    },
    "bodyMBold": {
      "fontFamily": "Inter",
      "fontSize": "16",
      "fontWeight": "600",
      "lineHeight": "24",
    },
    "bodyS": {
      "fontFamily": "Inter",
      "fontSize": "14",
      "fontWeight": "400",
      "lineHeight": "16",
    },
    "bodySBold": {
      "fontFamily": "Inter",
      "fontSize": "14",
      "fontWeight": "600",
      "lineHeight": "16",
    },
    "caption": {
      "fontFamily": "Inter",
      "fontSize": "12",
      "fontWeight": "400",
      "lineHeight": "12",
    },
    "code": {
      "fontFamily": "Menlo",
      "fontSize": "14",
      "fontWeight": "400",
      "lineHeight": "24",
    },
    "h1": {
      "fontFamily": "Inter",
      "fontSize": "24",
      "fontWeight": "600",
      "lineHeight": "32",
    },
    "h2": {
      "fontFamily": "Inter",
      "fontSize": "20",
      "fontWeight": "600",
      "lineHeight": "32",
    },
    "h3": {
      "fontFamily": "Inter",
      "fontSize": "18",
      "fontWeight": "600",
      "lineHeight": "24",
    },
    "h4": {
      "fontFamily": "Inter",
      "fontSize": "16",
      "fontWeight": "600",
      "lineHeight": "24",
    },
    "link": {
      "fontFamily": "Inter",
      "fontSize": "16",
      "fontWeight": "400",
      "lineHeight": "24",
      "textDecoration": "underline",
    },
  },
}
`;
